File: src/main/java/util/Digit.java
package util;

import java.io.Serializable;
import java.util.ArrayList;

public class Digit implements Serializable {
	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;

	public enum VALUE_TYPE{
		EMPTY,
		CLUE,
		NUMBER,
		GUESS
	};
	private ArrayList<Integer> value;
	private VALUE_TYPE type;
	
	public Digit(int value, VALUE_TYPE type){
		this.value = new ArrayList<>();
		this.value.add((Integer)value);
		this.type = type;
	}
	
	public Digit(){
		this.value = new ArrayList<>();
		this.type = VALUE_TYPE.EMPTY;
	}
	
	public VALUE_TYPE getType(){
		return this.type;
	}
	public void setType(VALUE_TYPE toSet){
		this.type = toSet;
	}
	public int get(int at){
		return value.get(at);
	}
	public int size(){
		return value.size();
	}
	public Object clone(){
		return value.clone();
	}
	public ArrayList<Integer> getValue(){
		return value;
	}
	public void setValue(ArrayList<Integer> toSet){
		value = toSet;
	}
	public void add(int toAdd, VALUE_TYPE type){
		if(this.type != VALUE_TYPE.CLUE){
			if(type == VALUE_TYPE.NUMBER){
				value.clear();
				value.add(toAdd);
				this.type = type;
			}
			else if(type == VALUE_TYPE.GUESS){
				try{
					value.remove((Integer)toAdd);
				}catch(IndexOutOfBoundsException e){}
				value.add(toAdd);
				this.type = type;
			}
		}
		java.util.Collections.sort(value);
	}
	public void remove(int toRemove){
		if(type != VALUE_TYPE.CLUE){
			try{
				value.remove((Integer)toRemove);
			}catch(NullPointerException e){}
			if(value.isEmpty())
				type = VALUE_TYPE.EMPTY;
		}
	}
	public void remove(){
		if(type != VALUE_TYPE.CLUE){
			value.clear();
			type = VALUE_TYPE.EMPTY;
		}
	}
	public void addClue(int toAdd){
		format();
		value.add(toAdd);
		type = VALUE_TYPE.CLUE;
	}
	public void format(){
		value.clear();
		type = VALUE_TYPE.EMPTY;
	}
	public void intersection(ArrayList<Integer> toUnite){
		for(int i=0; i<value.size(); ){
			if(!toUnite.contains(value.get(i))){
				value.remove(i);
			}
			else
				i++;
		}
	}
}

File: src/main/kotlin/util/Digit.kt
package util

import java.io.Serializable
import java.util.*

class Digit: Serializable {
    var value: MutableList<Int> = mutableListOf()
    var type: ValueType

    enum class ValueType {
        EMPTY,
        CLUE,
        NUMBER,
        GUESS
    }

    constructor(value: Int, type: ValueType) {
        this.value.add(value)
        this.type = type
    }

    constructor() {
        this.value = ArrayList()
        this.type = ValueType.EMPTY
    }

    operator fun get(at: Int): Int {
        return value[at]
    }

    fun size(): Int {
        return value.size
    }

    fun clone(): MutableList<Int> {
        return value
    }

    fun add(toAdd: Int, type: ValueType) {
        if (this.type != ValueType.CLUE) {
            if (type == ValueType.NUMBER) {
                value.clear()
                value.add(toAdd)
                this.type = type
            } else if (type == ValueType.GUESS) {
                try {
                    value.remove(toAdd)
                } catch (e: IndexOutOfBoundsException) {
                }

                value.add(toAdd)
                this.type = type
            }
        }
        value.sort()
    }

    fun remove(toRemove: Int) {
        if (type != ValueType.CLUE) {
            try {
                value.remove(toRemove)
            } catch (e: NullPointerException) {
            }

            if (value.isEmpty())
                type = ValueType.EMPTY
        }
    }

    fun remove() {
        if (type != ValueType.CLUE) {
            value.clear()
            type = ValueType.EMPTY
        }
    }

    fun addClue(toAdd: Int) {
        format()
        value.add(toAdd)
        type = ValueType.CLUE
    }

    fun format() {
        value.clear()
        type = ValueType.EMPTY
    }

    fun intersection(toUnite: MutableList<Int>) {
        var i = 0
        while (i < value.size) {
            if (!toUnite.contains(value[i])) {
                value.removeAt(i)
            } else
                i++
        }
    }

    companion object {
        /**
         *
         */
        private const val serialVersionUID = 1L
    }
}
